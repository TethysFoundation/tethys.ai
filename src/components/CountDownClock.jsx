import React from 'react';
import PropTypes from 'prop-types';
import { forbidExtraProps } from 'airbnb-prop-types';
import { I18n } from 'react-i18next';
import PieCounter from './PieCounter';
import styles from '../assets/stylesheets/count_down_clock.css';

function formatDate(date) {
  const month = padDatePart(date.getMonth() + 1);
  const day = padDatePart(date.getDate());
  const year = date.getFullYear();
  return `${month}/${day}/${year}`;
}

function padDatePart(d) {
  const str = d.toString();
  return str.length === 1 ? `0${str}` : str;
}

export default class CountDownClock extends React.Component {
  static propTypes = forbidExtraProps({
    endDate: PropTypes.instanceOf(Date).isRequired,
  });

  constructor(props) {
    super(props);
    this.endTime = props.endDate.getTime();
  }

  getDaysRemaining = () => {
    const currentTime = new Date().getTime();
    const msRemaining = Math.max(0, this.endTime - currentTime);
    return msRemaining / (1000 * 60 * 60 * 24);
  };

  getHoursRemaining = () => {
    const currentTime = new Date().getTime();
    const msRemaining = Math.max(0, this.endTime - currentTime);
    return (msRemaining / (1000 * 60 * 60)) % 24;
  };

  getMinutesRemaining = () => {
    const currentTime = new Date().getTime();
    const msRemaining = Math.max(0, this.endTime - currentTime);
    return (msRemaining / 1000 / 60) % 60;
  };

  getMillisRemaining = () => {
    const currentTime = new Date().getTime();
    const msRemaining = (this.endTime - currentTime) % 60000;
    return Math.max(0, msRemaining);
  };

  displayDaysRemaining = () => Math.floor(this.getDaysRemaining());

  displayMinutesRemaining = () => Math.floor(this.getMinutesRemaining());

  displaySecondsRemaining = () => Math.floor(this.getMillisRemaining() / 1000);

  displayHoursRemaining = () => Math.floor(this.getHoursRemaining());

  render() {
    const { endDate } = this.props;

    return (
      <I18n>
        {t => (
          <div className={styles.container}>
            <h2 data-testid="heading" className={styles.heading}>
              {t('countDown.heading', { endDate: formatDate(endDate) })}
            </h2>

            <div data-testid="counter-container" className={styles.countDown}>
              <PieCounter
                getValue={this.getDaysRemaining}
                displayValue={this.displayDaysRemaining}
                max={75}
                label={t('countDown.days')}
              />
              <PieCounter
                getValue={this.getHoursRemaining}
                displayValue={this.displayHoursRemaining}
                max={24}
                label={t('countDown.hours')}
              />
              <PieCounter
                getValue={this.getMinutesRemaining}
                displayValue={this.displayMinutesRemaining}
                max={60}
                label={t('countDown.minutes')}
              />
              <PieCounter
                getValue={this.getMillisRemaining}
                displayValue={this.displaySecondsRemaining}
                max={60000}
                label={t('countDown.seconds')}
              />
            </div>
          </div>
        )}
      </I18n>
    );
  }
}
