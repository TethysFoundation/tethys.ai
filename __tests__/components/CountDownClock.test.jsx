import React from 'react';
import { render } from 'react-testing-library';
import 'dom-testing-library/extend-expect';
import tk from 'timekeeper';
import CountDownClock from '../../src/components/CountDownClock';

describe('CountDownClock', () => {
  const endDate = new Date(new Date().getTime() + (15 * 24 * 60 * 60 * 1000));

  it('renders correctly with a future endDate', () => {
    const { getByTestId } = render(<CountDownClock endDate={endDate} />);

    expect(getByTestId('days-value')).toHaveTextContent('14');
    expect(getByTestId('hours-value')).toHaveTextContent('23');
    expect(getByTestId('minutes-value')).toHaveTextContent('59');
    expect(getByTestId('seconds-value')).toHaveTextContent('59');
  });

  it('renders all zeros when the endDate has been reached', () => {
    const time = new Date(endDate.getTime() + 1000);
    tk.freeze(time);

    const { getByTestId } = render(<CountDownClock endDate={endDate} />);

    expect(getByTestId('days-value')).toHaveTextContent('0');
    expect(getByTestId('hours-value')).toHaveTextContent('0');
    expect(getByTestId('minutes-value')).toHaveTextContent('0');
    expect(getByTestId('seconds-value')).toHaveTextContent('0');
  });
});
