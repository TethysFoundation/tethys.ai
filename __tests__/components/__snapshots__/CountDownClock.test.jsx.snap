// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountDownClock renders all zeros when the endDate has been reached 1`] = `
<div
  class="countDown"
  data-testid="counter-container"
>
  <div>
    <span>
      0
    </span>
    <span>
      0
    </span>
    <span>
      30
    </span>
    <span>
      Days
    </span>
  </div>
  <div>
    <span>
      0
    </span>
    <span>
      0
    </span>
    <span>
      24
    </span>
    <span>
      Hours
    </span>
  </div>
  <div>
    <span>
      0
    </span>
    <span>
      0
    </span>
    <span>
      60
    </span>
    <span>
      Minutes
    </span>
  </div>
  <div>
    <span>
      0
    </span>
    <span>
      0
    </span>
    <span>
      60000
    </span>
    <span>
      Seconds
    </span>
  </div>
</div>
`;

exports[`CountDownClock renders correctly with a future endDate 1`] = `
<div
  class="countDown"
  data-testid="counter-container"
>
  <div>
    <span>
      14.999988425925926
    </span>
    <span>
      14
    </span>
    <span>
      30
    </span>
    <span>
      Days
    </span>
  </div>
  <div>
    <span>
      23.999722222222204
    </span>
    <span>
      23
    </span>
    <span>
      24
    </span>
    <span>
      Hours
    </span>
  </div>
  <div>
    <span>
      59.983333333333576
    </span>
    <span>
      59
    </span>
    <span>
      60
    </span>
    <span>
      Minutes
    </span>
  </div>
  <div>
    <span>
      59000
    </span>
    <span>
      59
    </span>
    <span>
      60000
    </span>
    <span>
      Seconds
    </span>
  </div>
</div>
`;
